### ISSUE TITLE
CorrectForClockSkew drops local span

### ISSUE THREAD

**Issue by yurishkuro** [openzipkin/zipkin#945]
The attached trace has a local span called `get_root`. This span gets dropped from the trace by `CorrectForClockSkew`.  

[trace.json.txt](https://github.com/openzipkin/zipkin/files/118505/trace.json.txt)

The following diff seems to fix it, but I would like a second opinion about what's the best approach:

```
around @ -249,11 +262,25


+      // Local spans may have no annotations, so above calculation may produce None.
+      // But if the timestamp is already defined, we just need to adjust it by the skew.
+      val lcTimestamp: Option[Long] = span.timestamp.flatMap {
+        ts =>
+          span.binaryAnnotations.find(_.key == Constants.LocalComponent).map {
+            anno => anno.host match {
+              case Some(host) if isHost(host, anno.key) => ts - clockSkew.skew
+              case _ => ts
+            }
+          }
+      }
+
       new SpanTreeEntry(span.copy(
-        timestamp = firstOption,
-        duration = duration,
+        timestamp = lcTimestamp orElse firstOption,
+        duration = span.duration orElse duration,
         annotations = annotations.sorted), spanTree.children)
```

**Comment by yurishkuro** [comment]
Incidentally, there may be another issue (maybe with our instrumentation), because _all_ spans in this trace are actually from the same host, so the local span should have `skew=0`, which would've caused the adjustment logic to be skipped, yet that didn't happen. I haven't investigated yet why it finds a non-zero skew (most likely because of the parent spans).

**Comment by codefromthecrypt** [comment]
logic sounds sensible. ack that it should have skipped (vs dropping the local spans) in the first place

**Comment by yurishkuro** [comment]
@adriancole currently the change is limited to LC spans. It breaks one test in the span store spec that expects span-reported duration to be overwritten by the adjuster.

Should this actually apply to ALL spans? I.e. I can't think of a good reason for the adjuster to _ever_ override the span's timestamp / duration based on the annotations, it should only be allowed to shift the timestamp by the skew.

**Comment by codefromthecrypt** [comment]
> @adriancole currently the change is limited to LC spans. It breaks one
> test in the span store spec that expects span-reported duration to be
> overwritten by the adjuster.
> At first thought, I don't think duration should change, either... Either
> there's a non-intuitive reason why duration is changing, or there's a bug
> in the adjuster.
> 
> Should this actually apply to ALL spans? I.e. I can't think of a good
> reason for the adjuster to ever override the span's timestamp / duration
> based on the annotations, it should only be allowed to shift the timestamp
> by the skew.
> I don't quite understand what you mean, but I want to. I agree that
> adjusting the duration sounds curious. Let's place duration aside for a sec.

Are there conditions where the adjuster is incorrectly adjusting/overriding
the timestamp?

**Comment by codefromthecrypt** [comment]
Wait.. If there is clock skew, where one span should be shifted right, its
parent duration would be longer. I suppose what you are saying is that we
should only set the parent's duration if it was derived by annotations.

The problem is that there's a separate function to set duration from
annotations. If we want to make sure duration is set properly, the clock
skew correction has to happen before missing span.duration is filled in.

**Comment by codefromthecrypt** [comment]
ok just looked at the code.. seems clock skew is applied before span.timestamp, duration is derived from annotations.

So, to make sure we don't botch instrumentation-assigned duration, we need to do the following..

First, change the existing [clock skew test](https://github.com/openzipkin/zipkin/blob/master/zipkin-common/src/test/scala/com/twitter/zipkin/storage/SpanStoreSpec.scala#L368) to not set timestamp, duration on the parent and comment why. Basically, by leaving these unset, we can see if skewed, annotation-derived duration is set properly. The test impl should be the same.

Next, add a skew test where a parent's duration is set explicitly, but children are derived from annotations. One way to do this is to take the spans from the above test and make them a child of a span with an explicitly-set duration. In this case the parent will have its timestamp shifted, so that children happen after it, but the parent's duration is left alone. The child tree should be affected the same as the existing test.

@yurishkuro does ^^ capture your concern?

**Comment by yurishkuro** [comment]
@adriancole yes, sounds like the right approach.

**Comment by codefromthecrypt** [comment]
opened https://github.com/openzipkin/zipkin/issues/955 on the other topic

**Comment by codefromthecrypt** [comment]
closed via #954
